### PR TITLE
feat: show pending legal hold request and approve it [WPB-4393]

### PR DESCRIPTION
**Pull request by saleniuk**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-4393" title="WPB-4393" target="_blank">StoryWPB-4393</a>  [Android] Receive a Legal hold request
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
    - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
    - [x] contains a reference JIRA issue number like `SQPIT-764`
    - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
    - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

The logic in view models to show the bar informing about pending legal hold request and a dialog with the option to approve this request.

#### Test Coverage (Optional)

- [x] I have added automated test to this contribution

#### How to Test

As a team admin request the legal hold for the user, then this user should see this bar and dialog when he/she opens the app.

### Attachments (Optional)

https://github.com/wireapp/wire-android/assets/30429749/eeae3710-c825-43b0-aebc-f0e562a9f711

----
#### PR Post Submission Checklist for internal contributors (Optional)

- [x] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

- [x] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: app/src/main/kotlin/com/wire/android/ui/common/topappbar/CommonTopAppBarViewModel.kt [hunk 1]
```diff
@@ -77,6 +78,16 @@ class CommonTopAppBarViewModel @Inject constructor(
         }
     }
 
+    private fun legalHoldStatusFlow(userId: UserId) = coreLogic.sessionScope(userId) {
+        observeLegalHoldRequest() // TODO combine with legal host status
```
Comment (ohassine) [inline app/src/main/kotlin/com/wire/android/ui/common/topappbar/CommonTopAppBarViewModel.kt]: I think a typo here, you mean legal hold not host

File: app/src/main/kotlin/com/wire/android/ui/common/topappbar/CommonTopAppBarViewModel.kt [hunk 2]
```diff
@@ -77,6 +78,16 @@ class CommonTopAppBarViewModel @Inject constructor(
...
+            .map { legalHoldRequestResult ->
+                when (legalHoldRequestResult) {
+                    is ObserveLegalHoldRequestUseCaseResult.ObserveLegalHoldRequestAvailable -> LegalHoldStatus.PENDING
```
Comment (ohassine) [inline app/src/main/kotlin/com/wire/android/ui/common/topappbar/CommonTopAppBarViewModel.kt:85]: We may need to rename `ObserveLegalHoldRequestAvailable` to just `LegalHoldRequestAvailable` in kalium

File: app/src/main/kotlin/com/wire/android/ui/common/topappbar/CommonTopAppBarViewModel.kt [hunk 3]
```diff
@@ -77,6 +78,16 @@ class CommonTopAppBarViewModel @Inject constructor(
...
+                    else -> LegalHoldStatus.NO_CONSENT
```
Comment (ohassine) [inline app/src/main/kotlin/com/wire/android/ui/common/topappbar/CommonTopAppBarViewModel.kt]: Shouldn't be no_consent the same as pending ?  we display the pending banner of legal hold

File: app/src/main/kotlin/com/wire/android/ui/common/topappbar/CommonTopAppBarViewModel.kt [hunk 4]
```diff
@@ -148,7 +160,10 @@ class CommonTopAppBarViewModel @Inject constructor(
         LegalHoldStatus.DISABLED,
         LegalHoldStatus.NO_CONSENT -> LegalHoldUIState.None
     }.let { legalHoldUIState ->
-        if (currentScreen is CurrentScreen.AuthRelated) LegalHoldUIState.None
+        if (currentScreen is CurrentScreen.AuthRelated
+            || currentScreen is CurrentScreen.OngoingCallScreen
+            || currentScreen is CurrentScreen.IncomingCallScreen
```
Comment (ohassine) [inline app/src/main/kotlin/com/wire/android/ui/common/topappbar/CommonTopAppBarViewModel.kt]: I think we need to add `InitiatingCallScreen`

File: app/src/main/kotlin/com/wire/android/ui/common/topappbar/CommonTopAppBarViewModel.kt (hunk 1 above)
Comment (saleniuk) [reply app/src/main/kotlin/com/wire/android/ui/common/topappbar/CommonTopAppBarViewModel.kt]: in my defence, both words are similar 😂 
fixed

File: app/src/main/kotlin/com/wire/android/ui/common/topappbar/CommonTopAppBarViewModel.kt (hunk 2 above)
Comment (saleniuk) [reply app/src/main/kotlin/com/wire/android/ui/common/topappbar/CommonTopAppBarViewModel.kt:85]: Yeah, but it's just for better readability, so I won't raise a kalium PR just for that and will try to remember to fix that in the next kalium PR

File: app/src/main/kotlin/com/wire/android/ui/common/topappbar/CommonTopAppBarViewModel.kt (hunk 3 above)
Comment (saleniuk) [reply app/src/main/kotlin/com/wire/android/ui/common/topappbar/CommonTopAppBarViewModel.kt]: Right, first I did this and then started reading about this "consent", so it should be `DISABLED` and I changed that, and about "consent" - according to this: https://wearezeta.atlassian.net/wiki/spaces/PAD/pages/438894769/Legal+hold+on+Cloud+-+implemented+solution#Consent
"real" consent is marked as "future work" and for now, there is no way in the UI for the user to grant consent, we don't have UI or UX flow specified for that, instead, the "implicit" consent is given by the team administrator directly in the DB. In this specific case, for showing or hiding legal hold bar, we can just ignore `NO_CONSENT`.

File: app/src/main/kotlin/com/wire/android/ui/common/topappbar/CommonTopAppBarViewModel.kt (hunk 4 above)
Comment (saleniuk) [reply app/src/main/kotlin/com/wire/android/ui/common/topappbar/CommonTopAppBarViewModel.kt]: Sure, I added that, thanks!

**Review by ohassine — APPROVED** [review]

**Review by alexandreferris — APPROVED** [review]

**Pull request merged**